autolab/Autolab
# Overhaul LTI Configuration Settings, Remove dependency on lti_settings.yml

This PR moves LTI configuration settings out of "lti_settings.yml" and removes the requirement to have this file in order to run Autolab. Instead, a new LTI Configuration page and controller has been created to handle making and updating the LTI configuration. Platform and Tool JWK files are also no longer are inputted as PEM, but as JSON files.

Fixes #1677 

In the future, there should be tests for `lti_launch_controller` and `lti_nrps_controller`, as well as an easier way to import LTI configuration information, but as of now that is out of scope for this PR. Also, the `lti_config/index` page will get added to the admins dropdown in a future PR as well.

## Description

- create `lti_config_controller`
  - view of `lti_config_controller#index` to view current LTI configuration, modify settings
![Screen Shot 2023-02-03 at 9 48 15](https://user-images.githubusercontent.com/25730111/216632889-a840f4e2-b3e3-449f-8746-e0891c2b0519.png)
  - `lti_config_controller#update_config` takes form input, validates, updates configuration and uploads JWK files
    - uploading tool JWK file, platform JWK file not necessary if they already exist
    - Autolab will use platform JWK URL over platform JWK JSON file
    - User is not able to view tool JWK once it is uploaded, only overwrite (as it is a private key)
    - `lti_config_controller` will create `lti_config.yml` to store LTI configuration, as well as `lti_tool_jwk.json`, and `lti_platform_jwk.json` in `/config`, but can be modified in `config/environment/production.rb` if a different location is desired
 - Replace `lti_settings` with `config.lti_config_location` to point to directory where lti_config files are stored, and can be environment specific, also means Autolab can run without a LTI config
 - Create unit tests for `lti_config_controller` under `spec/controllers/lti_config_controller.rb`, have 100% coverage of controller
 - Update `autolab.rake` to create a non-admin instructor and non-instructor course assistant for testing purposes, add controller macros to make use of these new users
 - Update `lti_nrps_controller` and `lti_launch_controller` to use new LTI configuration
 - Update `lti_launch_controller`and `lti_config_controller` to use JWK JSON files over PEM format variables, no more embedded JWK constants and can support more than RSA format JWKs
   
## Motivation and Context



#1677 raised that JWK parameters were being embedded as constants in the `lti_nrps_controller` for the Tool (Autolab's) JWK. This means that credentials wouldn't have the correct `kid` unless the controller code was modified. This also restricted Tool JWKs to use only RSA key-types. The tool JWK was also being stored as a PEM format in Autolab, which is inconvenient to set up.

## How Has This Been Tested?



Run unit tests:
```
RAILS_ENV=test bundle exec rails autolab:setup_test_env
rake spec SPEC=./spec/controllers/lti_config_controller_spec.rb
```
1. Go to `/lti_config/index` as an admin, and fill in the parameters **manually** using `lti_config.yml`, and `lti_tool_jwk.json`, found in the zip file below
[lti_config_files.zip](https://github.com/autolab/Autolab/files/10580209/lti_config_files.zip)
2. Once you fill in all parameters, see that config is successfully created and `lti_config.yml` and `lti_tool_jwk.json` show up in `/config`, and the config should be saved on reloading the page
3. Try not including any of the required parameters, see that either the form pre-validates and tells you to fill in the field or you get a flash[:error]
![Screen Shot 2023-02-03 at 10 25 21](https://user-images.githubusercontent.com/25730111/216641261-3d683134-3731-4ce0-957d-92ae064294af.png)

Using a properly setup config, run through the LTI launch / LTI NRPS workflow to ensure it still works:

- Go to https://lti-ri.imsglobal.org/platforms/3644/resource_links/74024/rosters
- Select a user for launch (OIDC login)
![Screen Shot 2023-02-03 at 10 28 23](https://user-images.githubusercontent.com/25730111/216641987-a96e3e69-143c-4024-8beb-229da8001ac8.png)
- Click 'launch resource link', then click 'Post request' or 'Get request'
![Screen Shot 2023-02-03 at 10 28 59](https://user-images.githubusercontent.com/25730111/216642150-8d80d072-6ef5-4f17-840e-ec726e008498.png)
- Click 'Launch Resource link' and then 'Perform Launch'
- If you are prompted to log into Autolab, perform the login, then run through the process again
- You should then be redirected back to Autolab, where you can choose a course:

![image](https://user-images.githubusercontent.com/25730111/216642618-d626c7e7-3dfc-432e-ae3d-423a31702cdd.png)

- Go to 'Manage Course' for the Course you just linked with LTI, and then click 'Manage Users'
- Click on the refresh button
![Screen Shot 2023-02-03 at 10 31 58](https://user-images.githubusercontent.com/25730111/216642780-7bd7accb-c1f5-4963-a090-9094bd81f579.png)
- See that no errors appear

## Types of changes


## Checklist:

( will do later )
